### PR TITLE
[aptos-vm] init_module & state value metadata for V2 loader

### PR DESCRIPTION
**Pull request by georgemitenkov**
## Description

Two changes:
1. Beautiful hack thanks to @zekun000 to support init module without index struct name cache flushing.
2. Added API for `AptosModuleStorage` to query state value metadata.
Note: tests 2will come when things are connected e2e. 

Kept them as a separate PR for simplicity. This way it is easier to address all `TODO(loader_v2)`s... 

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by georgemitenkov** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14341?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14341** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14341?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14206** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14184** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14184?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#14075](https://github.com/aptos-labs/aptos-core/pull/14075) <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14075?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#14185** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @georgemitenkov and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: aptos-move/aptos-vm-types/src/module_and_script_storage/module_storage.rs [hunk 1]
```diff
@@ -1,9 +1,29 @@
...
+use move_core_types::{account_address::AccountAddress, identifier::IdentStr};
 use move_vm_runtime::{DummyCodeStorage, ModuleStorage};
 
 /// Represents module storage used by the Aptos blockchain.
-pub trait AptosModuleStorage: ModuleStorage {}
+pub trait AptosModuleStorage: ModuleStorage {
+    /// Returns the state value metadata of an associated with this module. The
+    /// error is returned if there is a storage error. If the module does not exist,
+    /// `None` is returned.
+    fn fetch_state_value_metadata(
```
Comment (gelash) [inline aptos-move/aptos-vm-types/src/module_and_script_storage/module_storage.rs:14]: how do we make sure to cleanly separate the functionality and apis vs the executor view traits? do the module interfaces just move here?

Comment (georgemitenkov) [reply aptos-move/aptos-vm-types/src/module_and_script_storage/module_storage.rs:14]: Yes! `ModuleView` is no longer needed, we have all the traits here, so this essentially merges executor traits into here. The reason is that we do not really need an adapter for modules, and we want to have a separate flow from data ideally. Block-STM will implement `AptosModuleStorage` directly.

Block-STM is generic, but I also previously added a trait method to convert addresses, module names into generic `K: ModulePath` (state keys) so we are good there as well.

File: aptos-move/aptos-vm/src/move_vm_ext/write_op_converter.rs [hunk 2]
```diff
@@ -186,8 +185,15 @@ impl<'r> WriteOpConverter<'r> {
                 Op::New(bytes)
             };
 
-            // TODO(loader_v2): Query state value metadata from module storage.
-            let write_op = self.convert_module(&state_key, op, false)?;
+            let state_value_metadata = module_storage.fetch_state_value_metadata(addr, name)?;
```
Comment (gelash) [inline aptos-move/aptos-vm/src/move_vm_ext/write_op_converter.rs:188]: Let's test this beautiful method, now that we purely depend on module storage interfaces. We can make sure new / modify cases work with the correct metadata.

Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/write_op_converter.rs:188]: done

**Comment by georgemitenkov** [comment]
@gelash merging this because init module here is not correct anyway, and metadata is trivial + added tests and addressed your comments.

**Pull request merged**